Add segment sorter for data streams

mayya-sharipova opened the pull request:
It is beneficial to sort segments within a datastream's index 
by desc order of their max timestamp field, so
that the most recent (in terms of timestamp) segments
will be first.

This allows to speed up sort query on @timestamp desc field,
which is the most common type of query for datastreams,
as we are mostly concerned with the recent data.
This patch addressed this for writable indices.

Segments' sorter is different from index sorting. 
 An index sorter by itself is only concerned about the order of docs 
within an individual segment (and not how the segments are organized),
while the segment sorter is only used during search  and allows 
to start docs collection  with the "right" segment, 
so we can terminate the collection faster.

TODO:
- Should we also do this for frozen or readonly indices?  
Or they always supposed to have 1 segment?

mayya-sharipova committed:
Add default leaf sorter for data streams

It is beneficial to add leaf sorter for data streams
by desc order of their max timestamp field, so
that the most recent (in terms of timestamp) segments
will be first.
This allows to speed up sort query on @timestamp desc field,
which is the most common type of query for datastreams,
as we are mostly concerned with the recent data.
This patch addressed for writable indices.

This path also adds a new flag isDataStreamIndex
to IndexMetadata that shows if this index is a
part of datastream or not.

TODO:
- Should we also to this for frozen or readonly indices?

elasticmachine commented:
Pinging @elastic/es-search (Team:Search)

mayya-sharipova commented:
@elasticmachine run elasticsearch-ci/part-1

dakrone commented:
@mayya-sharipova I have a question because I am unfamiliar with this part of the code :) Is this leaf sorter the same as regular index sorting, or is this something that only takes effect on query or merging? Does this affect the indexing throughput the way the index-time sorting can?

mayya-sharipova commented:
@dakrone Thank for taking a look.  Answering your questions: 

>  Is this leaf sorter the same as regular index sorting, or is this something that only takes effect on query or merging? 

Segments' sorter is different from index sorting, and is only used during querying.  An index sorter by itself is only concerned about the order of docs  within an individual segment (and not how the segments are organized), while the segment sorter helps to start with docs collection during querying with the "right" segment, so we can terminate the collection faster once we collected to the top docs (the most recent docs in the case of datastreams).

> Does this affect the indexing throughput the way the index-time sorting can?

No, it doesn't.  Segment's sorter is only used during querying. 

I've also clarified the PR comment with this information.

hendrikmuhs commented:
Does sorting get always applied?

In the case of rollup and transform a composite agg is used and both read data in ascending order (date histogram), in addition both can not benefit from an early exit, but actually disable this explicitly.

That's why I am concerned about this change if it changes the behavior for _all_ cases.

jimczi commented on server/src/main/java/org/elasticsearch/cluster/metadata/IndexMetadata.java:
```diff
@@ -387,6 +389,7 @@ public static APIBlock readFrom(StreamInput input) throws IOException {
     private final ActiveShardCount waitForActiveShards;
     private final ImmutableOpenMap<String, RolloverInfo> rolloverInfos;
     private final boolean isSystem;
+    private final boolean isDataStreamIndex;
```
I don't think we should change the index metadata for this information. That would only work on data streams created after 7.15 unless I missed something ?
Checking if a concrete index is a data stream can be done dynamically when the `IndexShard` or `IndexService` is created ? You can access the index abstraction lookup with `clusterService.state().metadata().getIndicesLookup()`.

danhermann replied:
+1 to what @jimczi said here. Individual indices can be detached from and added to data streams in a variety of scenarios including snapshot restores the relatively new [Migrate to Data Stream API](https://www.elastic.co/guide/en/elasticsearch/reference/current/indices-migrate-to-data-stream.html). If such a flag were included in index metadata, code would need to be added in all such places to correct the index metadata and we'd prefer to avoid that.

mayya-sharipova replied:
@jimczi  Thank you for comment. I've started with a dynamic lookup `clusterService.state().metadata().getIndicesLookup()` like you suggested but it is not allowed to to access a cluster state while modifying cluster state (e.g. during index creation). There is a function `ClusterApplierService::assertNotCalledFromClusterStateApplier` that checks for it, and returns an error: "the applied cluster state is not yet available". I can investigate if it is possible to do this check somewhere earlier.

@danhermann Thanks for checking, and highlighting Migrate Data Stream API. I would think that this API would be quite rare to use, and most datastream indices would be created in a usual way.  But indeed if used, we would need to update  index metadata.

I will see if it is possible to do this check dynamically,  but as we build more code around datastreams it would be nice to have an easy way to check if a shard is a part of datastream, with APIs like:   `IndexMetadata::isDataStreamIndex` or `IndexShard::isDataStreamIndex`.

jimczi reviewed (commented):
I left a comment regarding the index metadata change. It should be possible to detect if an index is part of a datastream dynamically.

mayya-sharipova commented:
@hendrikmuhs Thanks for looking and raising your concerns. Answering your questions: 

> Does sorting get always applied?

> In the case of rollup and transform a composite agg is used and both read data in ascending order (date histogram), in addition both can not benefit from an early exit, but actually disable this explicitly.

Yes, it will be always applied to datastream indices. But this PR is about sorting of segments, in terms of which segment is better to start when doing document collection. It adds a tiny overhead, and for aggregations that collect all documents from all segments, it doesn't really matter from what segment to start.

I think a bigger concern for your use case could be https://github.com/elastic/elasticsearch/issues/71186, which proposes to add  sort on @timestamp by default to all searches on datastream indices. Please raise your very valid concerns on that issue.

mayya-sharipova committed:
Revert "Add default leaf sorter for data streams"

This reverts commit 53609871addeb9a7918192bb58debcabc561730f.
As we want to dynamically specify in an index belongs to data stream.

hendrikmuhs commented:
@mayya-sharipova 

Thanks for the response. It would be good to benchmark it. If you need help to set this up let me know. It seems tricky to benchmark, I might be wrong, but a lot of benchmarks force-merge to 1 segment before testing, we obviously don't want that behavior in this case.

I have a non-PR'd yet transform track which might be useful, however we don't need transform for this. The aggregation tracks might work as well.

mayya-sharipova committed:
Add segment sorter for data streams

It is beneficial to sort segments within a datastream's index
by desc order of their max timestamp field, so
that the most recent (in terms of timestamp) segments
will be first.

This allows to speed up sort query on @timestamp desc field,
which is the most common type of query for datastreams,
as we are mostly concerned with the recent data.
This patch addressed this for writable indices.

Segments' sorter is different from index sorting.
An index sorter by itself is only concerned about the order of docs
within an individual segment (and not how the segments are organized),
while the segment sorter is only used during search and allows
to start docs collection with the "right" segment,
so we can terminate the collection faster.

This PR adds a property to IndexShard `isDataStreamIndex` that
shows if a shard is a part of datastream.

mayya-sharipova commented on server/src/main/java/org/elasticsearch/cluster/metadata/DataStream.java:
```diff
@@ -36,6 +40,21 @@
+                        byte[] sortValue = points.getMaxPackedValue();
+                        return LongPoint.decodeDimension(sortValue, 0);
+                    }
+                } catch (IOException e) {
```
added in a9d44d97d793246644ecc955b8889aef5cf86738

mayya-sharipova replied:
@jimczi   I've opted to throw `IllegalStateException` instead of assertion.  This allows us to to return an arbitrary value as it was before.

I hope you are fine with the change.

mayya-sharipova commented:
@jimczi  Thanks for your review so far. In a9d44d97d793246644ecc955b8889aef5cf86738, I've added a data-stream check as you suggested; the check is done in `IndicesClusterStateService::createShard`, and now it will be a property of `IndexShard` object instead of a part of stored metadata.

There is an unresolved issue with migrate with data stream API,  as there is no closing and reopening of shard during migration, a migration to  data stream on already existing migrated indices will not be reflected. We have two options to address it:
1.  Decide that it doesn't matter. Migration happens rarely, this only affects speedups in search and on new indices it will be reflected.
2. @dnhatn suggested to apply leaf sorter not only on data stream indices, but on all indices that have `@timestamp` field. For this,  we can check `MapperService` in a similar way as it is done in `TimestampFieldMapperService`.

@jimczi  WDYT?

jpountz commented:
@hendrikmuhs Do I understand you correctly that benchmarking date histograms on `@timestamp` should be a good enough proxy to make sure that transforms are unaffected by this change? Mayya and I just discussed about doing this before merging this change.

jimczi commented:
> @dnhatn suggested to apply leaf sorter not only on data stream indices, but on all indices that have @timestamp field. For this, we can check MapperService in a similar way as it is done in TimestampFieldMapperService.

I like the idea. That'll probably require to move the timestamp metadata field in core but that makes sense to me.

hendrikmuhs commented:
> @hendrikmuhs Do I understand you correctly that benchmarking date histograms on `@timestamp` should be a good enough proxy to make sure that transforms are unaffected by this change? Mayya and I just discussed about doing this before merging this change.

Yes, that sound like a good plan to me. Thanks for looking into it!

dimitris-athanasiou commented:
Another use case that might be affected is scrolling through all docs in `ascending` order of a `date` field. This is what ML anomaly detection does when aggregations can't be used (which is many use cases). It might be worth also benchmarking if segment sorting slows this use case down significantly.

mayya-sharipova committed:
Merge remote-tracking branch 'upstream/master' into datastreams-leaf-sorter

mayya-sharipova committed:
IndexShard to use mappingLookup().isDataStreamTimestampFieldEnabled()
after merging of PR#75906

mayya-sharipova commented:
@elasticmachine run elasticsearch-ci/part-1

mayya-sharipova committed:
Merge remote-tracking branch 'upstream/master' into datastreams-leaf-sorter

mayya-sharipova committed:
Merge remote-tracking branch 'upstream/master' into datastreams-leaf-sorter

mayya-sharipova committed:
Check for datastream index only when MapperService is enabled

mayya-sharipova commented:
@elasticmachine test this please

mayya-sharipova committed:
Merge remote-tracking branch 'upstream/master' into datastreams-leaf-sorter

mayya-sharipova committed:
Temp

mayya-sharipova commented:
@elasticmachine test this please

mayya-sharipova commented:
I've ran benchmarks on [http_logs](https://github.com/elastic/rally-tracks/tree/master/http_logs) dataset (making http_logs a datastream).

Here are the results of the  99th percentile service time.
Baseline: master branch
Contender: this PR

```diff
  |                           Task |    Baseline |   Contender |     Diff |    Unit |
- |                          range |     9.64485 |     10.8926 |  1.24772 |      ms |
+ |            desc_sort_timestamp |     659.543 |       90.58 | -568.963 |      ms |
+ | desc_sort_with_after_timestamp |     929.417 |     833.383 | -96.0341 |      ms |
+ |             asc_sort_timestamp |     4.67752 |     4.40691 | -0.27061 |      ms |
- |  asc_sort_with_after_timestamp |     873.479 |     943.137 |  69.6586 |      ms |
- |                     hourly_agg |     2596.41 |     2675.71 |  79.3014 |      ms |
+ |                         scroll |     25056.4 |     22846.8 | -2209.57 |      ms |
```

- desc_sort now is performed much better, as we sort segment readers by desc date.
- asc_sort_with_after got  7% worse, should not be a concert as we have changes from Lucene to improve sort with after
-  hourly_agg got 3% worse, shouldn't be a big concern
- scroll got 9% better, not sure how to explain this improvement, as I modified rally-tracks to have scroll on `{"@timestamp" : "asc"`

jimczi commented on server/src/main/java/org/elasticsearch/cluster/metadata/DataStream.java:
```diff
@@ -36,6 +40,23 @@
+                    assert false : "Datastream index segment doesn't contain an expected " +
+                        DataStream.TimestampField.FIXED_TIMESTAMP_FIELD + " field!";
+                }
+                // this should not happen, as all data stream segments must contain @timestamp field
```
should we have an assert here too ?

mayya-sharipova replied:
@jimczi  I've opted to throw IllegalStateException instead of assertion. This allows us to to return an arbitrary value as it was before.

I hope you are fine with the change.

jimczi commented on server/src/main/java/org/elasticsearch/index/shard/IndexShard.java:
```diff
@@ -387,6 +389,7 @@ public boolean shouldCache(Query query) {
         persistMetadata(path, indexSettings, shardRouting, null, logger);
         this.useRetentionLeasesInPeerRecovery = replicationTracker.hasAllPeerRecoveryRetentionLeases();
         this.refreshPendingLocationListener = new RefreshPendingLocationListener();
+        this.isDataStreamIndex = mapperService == null ? false : mapperService.mappingLookup().isDataStreamTimestampFieldEnabled();
```
nice!

jimczi reviewed (approved):
LGTM, thanks for iterating on this.

mayya-sharipova committed:
Throw an exception when datastream doesn't contain

@timestamp field.

mayya-sharipova committed:
Merge remote-tracking branch 'upstream/master' into datastreams-leaf-sorter

mayya-sharipova committed:
Remove temp debugging

mayya-sharipova committed:
Change IllegalArgument to IllegatState exception

mayya-sharipova committed:
Return Long.MIN_VALUE for datastream sorter
if the segment doesn't contain documents.

mayya-sharipova commented:
@elasticmachine run elasticsearch-ci/part-2

mayya-sharipova committed:
Merge remote-tracking branch 'upstream/master' into datastreams-leaf-sorter

mayya-sharipova commented on server/src/main/java/org/elasticsearch/cluster/metadata/DataStream.java:
```diff
@@ -36,6 +40,25 @@
+                }
+                // this may happen if the segment contains only deleted documents
+                assert(r.numDocs() == 0);
+                return Long.MIN_VALUE;
```
@jimczi  It looks like we need to allow to have points to be `null` for cases when a segment contains only deleted documents.  Are you ok with the current implementation:

- Throw IllegalStateException if we can't access points
- Return `Long.MIN_VALUE` if a segment contains only deleted docs

jimczi replied:
We can also check the number of docs upfront ?:
````
if (r.numDocs() == 0) {
  return Long.MIN_VALUE;
}
````

mayya-sharipova replied:
Good suggestion, done in 4c5bddc72400e1a4af9da6292754ea7cf8c8ace4

mayya-sharipova committed:
Change assertion to if statement

mayya-sharipova commented:
@elasticmachine run elasticsearch-ci/bwc

mayya-sharipova committed:
Merge remote-tracking branch 'upstream/master' into datastreams-leaf-sorter
